### DISCUSSION TITLE
[1.1.236] Support DW Openpose preprocessor

### DISCUSSION THREAD

**Discussion by huchenlei** [Mikubill/sd-webui-controlnet#1863] [General]
See initial issue here: https://github.com/Mikubill/sd-webui-controlnet/issues/1855

DW Openpose preprocessor greatly improves the accuracy of openpose detection especially on hands. Here is a comparison used in our unittest:

Input Image:
![woman](https://github.com/Mikubill/sd-webui-controlnet/assets/20929282/918b8d6d-4e05-4609-9a1b-34a759edb7cd)

Openpose Full Output (Right hand missing):
![expected_woman_all_output](https://github.com/Mikubill/sd-webui-controlnet/assets/20929282/eb781c40-c46b-499e-b571-d7b84008943e)

DW Openpose Full Output:
![expected_woman_dw_all_output](https://github.com/Mikubill/sd-webui-controlnet/assets/20929282/1ac12389-f0a8-4515-a13f-eb58afb176fa)

Usage:
![dw_pose](https://github.com/Mikubill/sd-webui-controlnet/assets/20929282/58c5203d-46b6-42e1-abd9-16c334e07252)

With this pose detection accuracy improvements, we are hyped to start re-train the ControlNet openpose model with more accurate annotations. There is a proposal in DW Pose repository: https://github.com/IDEA-Research/DWPose/issues/2.

I would really want @lllyasviel to take the initiative for this retraining task, but he probably busy with other tasks. I am not familiar with ControlNet training, so I welcome suggestions and ideas on how should we proceed on this task with community support. You can find me on my discord server: https://discord.gg/GkaWcUat7R. Or just leave a comment here.

**Comment by huchenlei** [comment]
Note:

According to https://github.com/Mikubill/sd-webui-controlnet/issues/1864#issuecomment-1666262013, you need to install Microsoft Visual C++ (https://visualstudio.microsoft.com/visual-cpp-build-tools/) for some dependencies to build correctly.

**Reply by Erwin11** [reply]
run DW Openpose preprocessor  ,it shows [TypeError: 'NoneType' object is not iterable]
version:  ControlNet v1.1.237, SD v1.5.1,

full code:
```
Python 3.10.9 (tags/v3.10.9:1dd9be6, Dec  6 2022, 20:01:21) [MSC v.1934 64 bit (AMD64)]
Version: v1.5.1
Commit hash: 68f336bd994bed5442ad95bad6b6ad5564a5409a

current transparent-background 1.2.4

Installing sd-dynamic-prompts requirements.txt

Installing requirements for TemporalKit extension

Launching Web UI with arguments: --medvram --opt-sdp-no-mem-attention --api
No module 'xformers'. Proceeding without it.
[-] ADetailer initialized. version: 23.7.9, num models: 9
2023-08-05 08:10:39,593 - ControlNet - INFO - ControlNet v1.1.237
ControlNet preprocessor location: E:\VM\test_model\stable-diffusion-webui-1.4\stable-diffusion-webui\extensions\sd-webui-controlnet\annotator\downloads
2023-08-05 08:10:39,758 - ControlNet - INFO - ControlNet v1.1.237
sd-webui-prompt-all-in-one background API service started successfully.
Image Browser: ImageReward is not installed, cannot be used.
Loading weights [5493a0ec49] from E:\VM\test_model\stable-diffusion-webui-1.4\stable-diffusion-webui\models\Stable-diffusion\ACG\AOM3A1B_orangemixs.safetensors
Creating model from config: E:\VM\test_model\stable-diffusion-webui-1.4\stable-diffusion-webui\configs\v1-inference.yamlLatentDiffusion: Running in eps-prediction mode
DiffusionWrapper has 859.52 M params.
Loading VAE weights specified in settings: E:\VM\test_model\stable-diffusion-webui-1.4\stable-diffusion-webui\models\VAE\vae-ft-mse-840000-ema-pruned.safetensors
Applying attention optimization: sdp-no-mem... done.
Model loaded in 3.2s (load weights from disk: 0.3s, create model: 0.5s, apply weights to model: 0.6s, apply half(): 0.7s, load VAE: 0.2s, load textual inversion embeddings: 0.2s, calculate empty prompt: 0.7s).
Running on local URL:  http://127.0.0.1:7860

To create a public link, set `share=True` in `launch()`.
[AgentScheduler] Task queue is empty
[AgentScheduler] Registering APIs
Startup time: 87.5s (launcher: 38.9s, import torch: 4.0s, import gradio: 1.7s, setup paths: 1.8s, other imports: 1.7s, load scripts: 15.6s, create ui: 5.6s, gradio launch: 17.8s, app_started_callback: 0.3s).
2023-08-05 08:12:37,515 - ControlNet - INFO - Preview Resolution = 512
Loads checkpoint by http backend from path: https://download.openmmlab.com/mmdetection/v2.0/yolox/yolox_l_8x8_300e_coco/yolox_l_8x8_300e_coco_20211126_140236-d3bd2b23.pth
Loads checkpoint by local backend from path: E:\VM\test_model\stable-diffusion-webui-1.4\stable-diffusion-webui\extensions\sd-webui-controlnet\annotator\downloads\openpose\dw-ll_ucoco_384.pth
Traceback (most recent call last):
  File "E:\VM\test_model\stable-diffusion-webui-1.4\stable-diffusion-webui\venv\lib\site-packages\gradio\routes.py", line 422, in run_predict
    output = await app.get_blocks().process_api(
  File "E:\VM\test_model\stable-diffusion-webui-1.4\stable-diffusion-webui\venv\lib\site-packages\gradio\blocks.py", line 1323, in process_api
    result = await self.call_function(
  File "E:\VM\test_model\stable-diffusion-webui-1.4\stable-diffusion-webui\venv\lib\site-packages\gradio\blocks.py", line 1051, in call_function
    prediction = await anyio.to_thread.run_sync(
  File "E:\VM\test_model\stable-diffusion-webui-1.4\stable-diffusion-webui\venv\lib\site-packages\anyio\to_thread.py", line 33, in run_sync
    return await get_asynclib().run_sync_in_worker_thread(
  File "E:\VM\test_model\stable-diffusion-webui-1.4\stable-diffusion-webui\venv\lib\site-packages\anyio\_backends\_asyncio.py", line 877, in run_sync_in_worker_thread
    return await future
  File "E:\VM\test_model\stable-diffusion-webui-1.4\stable-diffusion-webui\venv\lib\site-packages\anyio\_backends\_asyncio.py", line 807, in run
    result = context.run(func, *args)
  File "E:\VM\test_model\stable-diffusion-webui-1.4\stable-diffusion-webui\extensions\sd-webui-controlnet\scripts\controlnet_ui\controlnet_ui_group.py", line 606, in run_annotator
    result, is_image = preprocessor(
  File "E:\VM\test_model\stable-diffusion-webui-1.4\stable-diffusion-webui\extensions\sd-webui-controlnet\scripts\utils.py", line 79, in decorated_func
    return cached_func(*args, **kwargs)
  File "E:\VM\test_model\stable-diffusion-webui-1.4\stable-diffusion-webui\extensions\sd-webui-controlnet\scripts\utils.py", line 67, in cached_func
    return func(*args, **kwargs)
  File "E:\VM\test_model\stable-diffusion-webui-1.4\stable-diffusion-webui\extensions\sd-webui-controlnet\scripts\global_state.py", line 35, in unified_preprocessor
    return preprocessor_modules[preprocessor_name](*args, **kwargs)
  File "E:\VM\test_model\stable-diffusion-webui-1.4\stable-diffusion-webui\extensions\sd-webui-controlnet\scripts\processor.py", line 265, in run_model
    return remove_pad(self.model_openpose(
  File "E:\VM\test_model\stable-diffusion-webui-1.4\stable-diffusion-webui\extensions\sd-webui-controlnet\annotator\openpose\__init__.py", line 360, in __call__
    poses = self.detect_poses_dw(oriImg)
  File "E:\VM\test_model\stable-diffusion-webui-1.4\stable-diffusion-webui\extensions\sd-webui-controlnet\annotator\openpose\__init__.py", line 337, in detect_poses_dw
    return Wholebody.format_result(keypoints_info)
  File "E:\VM\test_model\stable-diffusion-webui-1.4\stable-diffusion-webui\extensions\sd-webui-controlnet\annotator\openpose\wholebody.py", line 103, in format_result
    body = BodyResult(body_keypoints, total_score(body_keypoints), len(body_keypoints))
  File "E:\VM\test_model\stable-diffusion-webui-1.4\stable-diffusion-webui\extensions\sd-webui-controlnet\annotator\openpose\wholebody.py", line 93, in total_score
    return sum(keypoint.score for keypoint in keypoints if keypoint is not None)
TypeError: 'NoneType' object is not iterable

```

**Reply by huchenlei** [reply]
Fixed.

**Reply by Erwin11** [reply]
###  IT WORKS!  Thanks for your hard work!    small hands is  more accurate 

## DW Openpose
![dwpose](https://github.com/Mikubill/sd-webui-controlnet/assets/6300501/abbe3e3a-0051-4e1a-b745-2ad9503426dd)

## Openpose
![openpose ](https://github.com/Mikubill/sd-webui-controlnet/assets/6300501/2e311e2a-cd0e-4806-ac70-e9ddaac37418)

**Comment by huchenlei** [comment]
Due to #1875, the feature cannot be merged to main yet. I uploaded the version before reset to `dw_pose_reland` branch. If you still want to use the feature with a local compile of mmcv stuff, you can still switch to the `dw_pose_reland` branch for now while we work for a solution.

**Comment by p0mad** [comment]
@huchenlei 
Thanks fo this discussion, i have Controlnet 234 installed on my win pc.
is that enough to just checkout to the dw_pose_reland branch and its good to go? or it also needs some other tweaks?

SO as you said before the dw pth model was not intended to be used for auto1111. so i guess just git checkout would needed for this purpose, right? ( or any other better ways recommanded?)

Thanks

**Reply by C0nsumption** [reply]
Hey everyone. Where do I download the preprocessor?

EDIT:
forget it. Only fetched. forgot to checkout to the branch.

**Reply by huchenlei** [reply]
The feature has relanded on main. You don't need to checkout other branches now.

**Reply by C0nsumption** [reply]
I did for Vlad but all good now. Works fine on Auto1111 main

**Comment by vararth** [comment]
If I understand correctly, dw_openpose_full is the same as openpose_full? If not, how can I set up the dw_openpose_full?

**Comment by viccho** [comment]
Hi everyone.
i'm new here and new on SD. I nedd to use dw pose but i can't reach the goal. here is a printscreen , the preprocessor detect face, but the result did not follow the preproc. have you an idea of what i can do ? (my prompt is just " girl" )
![image](https://github.com/Mikubill/sd-webui-controlnet/assets/82603307/96289a04-7925-41f7-b0bd-e7764dc2f265)
thnak you :)

**Reply by zainmalik95** [reply]
@viccho were you ale to resolve that? I'm having the same issue

**Reply by huchenlei** [reply]
You are probably using the v1.0 ControlNet model. The v1.0 model does not have face support. You can download ControlNet v1.1 models here: https://huggingface.co/lllyasviel/ControlNet-v1-1/tree/main

**Reply by zainmalik95** [reply]
I'm using v1.1.410
<img width="1727" alt="image" src="https://github.com/Mikubill/sd-webui-controlnet/assets/46667557/4a250505-2cc1-4947-9565-bfdc3ec3e0df">

**Reply by zainmalik95** [reply]
I've tried the dw openpose as well but still not able to get the desired results.
<img width="852" alt="image" src="https://github.com/Mikubill/sd-webui-controlnet/assets/46667557/0f45a6e6-a99d-4576-b7f6-1efcc5a34f2d">

**Reply by huchenlei** [reply]
For your case, you are using SD1.5 ControlNet model to control SDXL base model.

Note: currently for SDXL controlnet models, none of them supports openpose full.

**Reply by zainmalik95** [reply]
Oh okay, which ones should i use? I actually have to cases to handle, one is change the face posing and the other is full body posing, what combination do you think will be the best for me? I'm actually learning so don't have the concepts clear as of now.
Thank you for the quick reply btw.

**Reply by huchenlei** [reply]
See https://github.com/Mikubill/sd-webui-controlnet/discussions/2039 for SDXL support details. Side note: updating the extension to v1.1.411 will enable `Edit` button feature that allows you to edit the pose in place.
